<html>

<head>
    <meta content="text/html; charset=UTF-8" http-equiv="content-type">
    <link rel="stylesheet" type="text/css" href="/lessons/styles.css">
</head>

<body class="c16 doc-content">
    <h1 class="c5 c7 c17">Procesarea erorilor și stiva apelurilor de funcții.</h1>
    <p class="c5 c7 c10"><span class="c3"></span></p>
    <p class="c5 c7"><span class="c8">Dac&#259; &icirc;n interiorul unei func&#539;ii apare o eroare care nu este
            tratat&#259;, interpretorul JavaScript iese din aceast&#259; func&#539;ie &#537;i revine la codul extern
            &icirc;n c&#259;utarea unui gestionar de erori. De exemplu:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.078b67b531be88cdbd59a305b4817b5a5fc0660a"></a><a
        id="t.0"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">A</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;func A
                        starts&quot;</span><span class="c1">);<br> &nbsp; &nbsp;callSomeFunc();<br> &nbsp;
                        &nbsp;</span><span class="c4">console</span><span class="c1">.log(</span><span
                        class="c9">&quot;func A ends&quot;</span><span class="c1">);<br>}<br></span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;program
                        ends&quot;</span><span class="c1">);</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Aici, func&#539;ia A apeleaz&#259; o func&#539;ie nedefinit&#259; callSomeFunc().
            Prin urmare, la executarea programului, c&acirc;nd func&#539;ia A este apelat&#259;, aceasta este
            &icirc;ntrerupt&#259;. Interpretorul iese &icirc;n codul extern &icirc;n c&#259;utarea unui gestionar de
            erori. Totu&#537;i, &icirc;n codul extern &icirc;n jurul apelului func&#539;iei A nu exist&#259; nicio
            structur&#259; try..catch definit&#259;, a&#537;a c&#259; &icirc;ntreaga execu&#539;ie a programului se va
            &icirc;ncheia &icirc;n mod necorespunz&#259;tor. Ie&#537;irea la consol&#259;:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c3">func A starts</span></p>
    <p class="c5 c7"><span class="c3">Uncaught ReferenceError: callSomeFunc is not defined</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; at A (index.html:11:2)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; at index.html:30:6</span></p>
    <p class="c5 c7"><span class="c8">Aceea&#537;i situa&#539;ie se aplic&#259; &#537;i pentru apelurile de func&#539;ii
            &icirc;nglobate. Dac&#259; &icirc;ntr-o func&#539;ie &icirc;nglobat&#259; apare o eroare &#537;i nu este
            tratat&#259;, interpretorul iese &icirc;n contextul extern - func&#539;ia exterioar&#259;. Dac&#259; eroarea
            nu este tratat&#259; nici &icirc;n acea func&#539;ie, interpretorul iese din nou &icirc;n contextul extern,
            p&acirc;n&#259; g&#259;se&#537;te un gestionar de erori. Dac&#259; eroarea nu este tratat&#259; &icirc;n
            nicio func&#539;ie sau &icirc;n codul global, programul se &icirc;ncheie &icirc;n mod necorespunz&#259;tor.
            De exemplu:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.4815ce220202f7bd0dda71e9ec5b501250610550"></a><a
        id="t.1"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">A</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;func A
                        starts&quot;</span><span class="c1">);<br> &nbsp; &nbsp;callSomeFunc();<br> &nbsp;
                        &nbsp;</span><span class="c4">console</span><span class="c1">.log(</span><span
                        class="c9">&quot;func A ends&quot;</span><span class="c1">);<br>}<br></span><span
                        class="c0">function</span><span class="c1">&nbsp;</span><span class="c4">B</span><span
                        class="c1">(){<br> &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;func B starts&quot;</span><span class="c1">);<br>
                        &nbsp; &nbsp;A();<br> &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;func B ends&quot;</span><span
                        class="c1">);<br>}<br></span><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">C</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;func C
                        starts&quot;</span><span class="c1">);<br> &nbsp; &nbsp;B();<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;func C
                        ends&quot;</span><span class="c1">);<br>}<br>C();<br></span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;program ends&quot;</span><span
                        class="c1">);</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Aici, func&#539;ia C este apelat&#259;, care apoi apeleaz&#259; func&#539;ia B,
            care la r&acirc;ndul s&#259;u apeleaz&#259; func&#539;ia A, iar aceasta din urm&#259; apeleaz&#259; o
            func&#539;ie inexistent&#259; callSomeFunc. Ca rezultat, &icirc;n func&#539;ia A apare o eroare. Deoarece
            func&#539;ia A nu trateaz&#259; eroarea, interpretorul caut&#259; &icirc;n mod secven&#539;ial un gestionar
            de erori &icirc;n func&#539;ia B, apoi &icirc;n func&#539;ia C &#537;i, &icirc;n final, &icirc;n contextul
            global. Totu&#537;i, deoarece niciunde eroarea nu este tratat&#259;, executarea programului se va
            &icirc;ncheia dup&#259; apari&#539;ia erorii:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c3">func C starts</span></p>
    <p class="c5 c7"><span class="c3">func B starts</span></p>
    <p class="c5 c7"><span class="c3">func A starts</span></p>
    <p class="c5 c7"><span class="c3">Uncaught ReferenceError: callSomeFunc is not defined</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; at A (index.html:11:2)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; at B (index.html:16:2)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; at C (index.html:27:2)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; at index.html:31:1</span></p>
    <p class="c5 c7"><span class="c8">Acum vom defini un gestionar de erori &icirc;n una dintre func&#539;ii, de
            exemplu, &icirc;n func&#539;ia C:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.d714e71e3446858083264569cb35eab2fbaf7068"></a><a
        id="t.2"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">A</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;func A
                        starts&quot;</span><span class="c1">);<br> &nbsp; &nbsp;callSomeFunc();<br> &nbsp;
                        &nbsp;</span><span class="c4">console</span><span class="c1">.log(</span><span
                        class="c9">&quot;func A ends&quot;</span><span class="c1">);<br>}<br></span><span
                        class="c0">function</span><span class="c1">&nbsp;</span><span class="c4">B</span><span
                        class="c1">(){<br> &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;func B starts&quot;</span><span class="c1">);<br>
                        &nbsp; &nbsp;A();<br> &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;func B ends&quot;</span><span
                        class="c1">);<br>}<br></span><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">C</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;func C
                        starts&quot;</span><span class="c1">);<br> &nbsp; &nbsp;</span><span class="c0">try</span><span
                        class="c1">{<br> &nbsp; &nbsp; &nbsp; &nbsp;B();<br> &nbsp; &nbsp;}<br> &nbsp;
                        &nbsp;</span><span class="c0">catch</span><span class="c1">{<br> &nbsp; &nbsp; &nbsp;
                        &nbsp;</span><span class="c4">console</span><span class="c1">.log(</span><span
                        class="c9">&quot;Error occured&quot;</span><span class="c1">);<br> &nbsp; &nbsp;}<br> &nbsp;
                        &nbsp;</span><span class="c4">console</span><span class="c1">.log(</span><span
                        class="c9">&quot;func C ends&quot;</span><span class="c1">);<br>}<br> <br>C();<br></span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;program
                        ends&quot;</span><span class="c1">);</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">&Icirc;n acest caz, c&acirc;nd apare o eroare &icirc;n func&#539;ia A,
            interpretorul caut&#259; un gestionar de erori &icirc;n acea func&#539;ie. Fiindc&#259; func&#539;ia A nu
            trateaz&#259; eroarea, interpretorul se uit&#259; &icirc;n func&#539;ia B &#537;i apoi &icirc;n func&#539;ia
            C. &Icirc;n func&#539;ia C se g&#259;se&#537;te un gestionar de erori, astfel c&#259; programul
            continu&#259; s&#259; ruleze dup&#259; tratarea erorii.</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c3">func C starts</span></p>
    <p class="c5 c7"><span class="c3">func B starts</span></p>
    <p class="c5 c7"><span class="c3">func A starts</span></p>
    <p class="c5 c7"><span class="c3">Error occured</span></p>
    <p class="c5 c7"><span class="c3">func C ends</span></p>
    <p class="c7 c18"><span class="c3">program ends</span></p>
    <p class="c5 c7"><span class="c8">&Icirc;n acest context, deoarece func&#539;iile A &#537;i B nu trateaz&#259;
            eroarea, ele nu mai continu&#259; execu&#539;ia.</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c19">Propagarea erorii &icirc;n sus pe stiva apelurilor de func&#539;ii.</span></p>
    <p class="c5 c7 c10"><span class="c3"></span></p>
    <p class="c5 c7"><span class="c8">Uneori, erorile sunt tratate &icirc;ntr-unul din apelurile &icirc;nglobate ale
            altor func&#539;ii. De exemplu, lu&#259;m urm&#259;toarea situa&#539;ie:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.bd858e71dd35380cc1528a92e4c0c3714b312c1c"></a><a
        id="t.3"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c2">// clas&#259; pentru o baz&#259; de date
                        condi&#539;ional&#259;</span><span class="c1"><br></span><span class="c0">class</span><span
                        class="c1">&nbsp;</span><span class="c4">Database</span><span class="c1">{<br> &nbsp;
                        &nbsp;</span><span class="c0">constructor</span><span class="c1">(){<br> &nbsp; &nbsp; &nbsp;
                        &nbsp;</span><span class="c0">this</span><span class="c1">.data = [</span><span
                        class="c9">&quot;Tom&quot;</span><span class="c1">, </span><span
                        class="c9">&quot;Sam&quot;</span><span class="c1">, </span><span
                        class="c9">&quot;Bob&quot;</span><span class="c1">];<br> &nbsp; &nbsp;}<br> &nbsp;
                        &nbsp;</span><span class="c2">// ob&#539;inerea datelor</span><span class="c1"><br> &nbsp;
                        &nbsp;getItem(index){ <br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c0">if</span><span
                        class="c1">(index &gt; </span><span class="c14">0</span><span class="c1">&nbsp;&amp;&amp; index
                        &lt; </span><span class="c0">this</span><span class="c1">.data.length)<br> &nbsp; &nbsp; &nbsp;
                        &nbsp; &nbsp; &nbsp;</span><span class="c0">return</span><span class="c1">&nbsp;</span><span
                        class="c0">this</span><span class="c1">.data[index];<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c0">else</span><span class="c1">&nbsp; &nbsp; </span><span class="c2">// dac&#259;
                        indexul este incorect - gener&#259;m o eroare</span><span class="c1"><br> &nbsp; &nbsp; &nbsp;
                        &nbsp; &nbsp; &nbsp;</span><span class="c0">throw</span><span class="c1">&nbsp;</span><span
                        class="c0">new</span><span class="c1">&nbsp;</span><span class="c4">RangeError</span><span
                        class="c1">(</span><span class="c9">&quot;Index invalid&quot;</span><span class="c1">);<br>
                        &nbsp; &nbsp;}<br> &nbsp; &nbsp;</span><span class="c2">// deschiderea bazei de date</span><span
                        class="c1"><br> &nbsp; &nbsp;open(){<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;Baza de date a fost
                        deschis&#259;&quot;</span><span class="c1">);<br> &nbsp; &nbsp;}<br> &nbsp; &nbsp;</span><span
                        class="c2">// &icirc;nchiderea bazei de date</span><span class="c1"><br> &nbsp;
                        &nbsp;close(){<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;Baza de date a fost
                        &icirc;nchis&#259;&quot;</span><span class="c1">);<br> &nbsp; &nbsp;}<br>}<br></span><span
                        class="c2">// func&#539;ie de &icirc;nveli&#537; pentru ob&#539;inerea unui obiect din baza de
                        date pe baza indexului</span><span class="c1"><br></span><span class="c0">function</span><span
                        class="c1">&nbsp;</span><span class="c4">get</span><span class="c1">(index) {<br> &nbsp; <br>
                        &nbsp; &nbsp;</span><span class="c0">const</span><span class="c1">&nbsp;db = </span><span
                        class="c0">new</span><span class="c1">&nbsp;Database();<br> &nbsp; &nbsp;db.open();
                        &nbsp;</span><span class="c2">// deschidem condi&#539;ionat baza de date</span><span
                        class="c1"><br> &nbsp; &nbsp;</span><span class="c0">try</span><span class="c1">&nbsp;{<br>
                        &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c0">return</span><span
                        class="c1">&nbsp;db.getItem(index); &nbsp; </span><span class="c2">// return&#259;m elementul
                        ob&#539;inut</span><span class="c1"><br> &nbsp; &nbsp;} </span><span
                        class="c0">catch</span><span class="c1">(err) {<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.error(err); </span><span class="c2">// dac&#259;
                        apare o eroare, o gestion&#259;m </span><span class="c1"><br> &nbsp; &nbsp;}<br> &nbsp;
                        &nbsp;db.close(); </span><span class="c2">// &icirc;nchidem condi&#539;ionat baza de
                        date</span><span class="c1"><br>}<br></span><span class="c2">// afi&#537;area
                        rezultatului</span><span class="c1"><br></span><span class="c0">function</span><span
                        class="c1">&nbsp;</span><span class="c4">printResult</span><span class="c1">(){<br> &nbsp;
                        &nbsp;</span><span class="c0">const</span><span class="c1">&nbsp;item = get(</span><span
                        class="c14">5</span><span class="c1">); &nbsp; &nbsp;</span><span class="c2">//
                        &icirc;ncerc&#259;m s&#259; ob&#539;inem elementul cu indexul 5</span><span class="c1"><br>
                        &nbsp; &nbsp;</span><span class="c4">console</span><span class="c1">.log(</span><span
                        class="c9">&quot;Ob&#539;inut din baza de date:&quot;</span><span class="c1">, item);
                    </span><span class="c2">// afi&#537;&#259;m elementul ob&#539;inut pe consol&#259;</span><span
                        class="c1"><br>}<br>printResult();</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Aici este definit&#259; o clas&#259; de baz&#259; a unei baze de date fictive
            numit&#259; Database. Pentru a interac&#539;iona cu datele, sunt definite trei func&#539;ii. Func&#539;iile
            simulate de deschidere &#537;i &icirc;nchidere a bazei de date - func&#539;iile open &#537;i close,
            respectiv - &#537;i func&#539;ia getItem, care returneaz&#259; un element specificat prin index din array-ul
            de date. Totu&#537;i, dac&#259; indexul furnizat nu este corect - mai mic dec&acirc;t 0 sau mai mare
            dec&acirc;t indexul permis - atunci se genereaz&#259; o eroare de tip RangeError.</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Mai trebuie men&#539;ionat c&#259; aceasta nu este singura abordare pentru
            implementarea unui astfel de func&#539;ional. Uneori, se folose&#537;te o alt&#259; strategie, &icirc;n
            care, &icirc;n cazul unui index sau ID incorect, func&#539;ia returneaz&#259; null, ceea ce &icirc;n anumite
            situa&#539;ii poate fi mai preferabil. &Icirc;ns&#259;, &icirc;n acest caz, pentru scopuri demonstrative, ne
            oprim la generarea unei excep&#539;ii &icirc;n cazul unui index incorect.</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Convenim c&#259;, pentru a interac&#539;iona cu baza de date, trebuie s&#259; o
            &quot;deschidem&quot; ini&#539;ial cu ajutorul metodei open, iar dup&#259; finalizarea lucrului cu ea,
            s&#259; o &quot;&icirc;nchidem&quot; cu ajutorul metodei close - o abordare destul de comun&#259; &icirc;n
            lucrul cu bazele de date. Totu&#537;i, pentru a abstractiza de toate aceste detalii, definim o func&#539;ie
            suplimentar&#259; numit&#259; get, care prime&#537;te un ID &#537;i acceseaz&#259; baza de date pentru a
            ob&#539;ine elementul corespunz&#259;tor ID-ului. &#536;i deoarece la apelul metodei getItem poate
            ap&#259;rea o eroare, aceasta este gestionat&#259; &icirc;ntr-o structur&#259; try..catch.</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.cac427d36f41739aac82d82ed3908172e4aebaf6"></a><a
        id="t.4"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">try</span><span class="c1">&nbsp;{<br> &nbsp; &nbsp;</span><span
                        class="c0">return</span><span class="c1">&nbsp;db.getItem(index); &nbsp; </span><span
                        class="c2">// return&#259;m elementul ob&#539;inut</span><span class="c1"><br>} </span><span
                        class="c0">catch</span><span class="c1">(err) {<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.error(err); </span><span class="c2">// dac&#259;
                        apare o eroare, o gestion&#259;m </span><span class="c1"><br>}</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Apoi, aceast&#259; func&#539;ie este apelat&#259; din alt&#259; func&#539;ie
            numit&#259; printResult, care afi&#537;eaz&#259; rezultatul ob&#539;inut:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.df84b9eecb6b0218ec3aad576d9e858ee95d3943"></a><a
        id="t.5"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">printResult</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c0">const</span><span class="c1">&nbsp;item = get(</span><span class="c14">5</span><span
                        class="c1">); &nbsp; &nbsp;</span><span class="c2">// &icirc;ncerc&#259;m s&#259; ob&#539;inem
                        elementul cu indexul 5</span><span class="c1"><br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;Ob&#539;inut din
                        baza de date:&quot;</span><span class="c1">, item); </span><span class="c2">// afi&#537;&#259;m
                        elementul ob&#539;inut pe consol&#259;</span><span class="c1"><br>}</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Dac&#259; privim la ie&#537;irea programului:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c3">Baza de date a fost deschis&#259;</span></p>
    <p class="c5 c7"><span class="c3">RangeError: Index invalid</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; la Database.getItem (index.html:19:19)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; la get (index.html:36:19)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; la printResult (index.html:44:18)</span></p>
    <p class="c5 c7"><span class="c3">&nbsp; &nbsp; la index.html:47:1</span></p>
    <p class="c5 c7"><span class="c3">get @ index.html:38</span></p>
    <p class="c5 c7"><span class="c3">printResult @ index.html:44</span></p>
    <p class="c5 c7"><span class="c3">(anonymous) @ index.html:47</span></p>
    <p class="c5 c7 c10"><span class="c3"></span></p>
    <p class="c5 c7 c10"><span class="c3"></span></p>
    <p class="c5 c7"><span class="c3">Baza de date a fost &icirc;nchis&#259;</span></p>
    <p class="c5 c7"><span class="c3">Ob&#539;inut din baza de date: undefined</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c15">Dac&#259; observ&#259;m c&#259; atunci c&acirc;nd se furnizeaz&#259; un index
            incorect, pe de o parte, eroarea este gestionat&#259;, dar pe de alt&#259; parte, &icirc;n func&#539;ia
        </span><span class="c6">printResult </span><span class="c15">ob&#539;inem un rezultat nedefinit (valoare
            undefined), iar noi nu avem idee despre ce s-a &icirc;nt&acirc;mplat. Aceast&#259; informa&#539;ie poate fi
            ob&#539;inut&#259; doar din ie&#537;irea consolei &icirc;n timpul execu&#539;iei. Cu toate acestea,
            func&#539;ia </span><span class="c6">get </span><span class="c8">nu trebuie neap&#259;rat s&#259;
            afi&#537;eze eroarea &icirc;n consol&#259;; eroarea poate fi gestionat&#259; &icirc;n alt mod. Prin urmare,
            exist&#259; nevoia de a transmite informa&#539;ii despre eroare &icirc;n sus pe stiva de apeluri a
            func&#539;iilor (p&acirc;n&#259; la func&#539;ia printResult).</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Pentru a realiza acest lucru, vom modifica codul pentru a propaga eroarea &icirc;n
            sus pe stiva de apeluri a func&#539;iilor:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.6e9a135ee6f13c66d0aa649ff9a9619dbef9d2ae"></a><a
        id="t.6"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">class</span><span class="c1">&nbsp;</span><span
                        class="c4">Database</span><span class="c1">{<br> &nbsp; &nbsp;</span><span
                        class="c0">constructor</span><span class="c1">(){<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c0">this</span><span class="c1">.data = [</span><span
                        class="c9">&quot;Tom&quot;</span><span class="c1">, </span><span
                        class="c9">&quot;Sam&quot;</span><span class="c1">, </span><span
                        class="c9">&quot;Bob&quot;</span><span class="c1">];<br> &nbsp; &nbsp;}<br> &nbsp; &nbsp;<br>
                        &nbsp; &nbsp;getItem(index){ <br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c0">if</span><span class="c1">(index &gt; </span><span class="c14">0</span><span
                        class="c1">&nbsp;&amp;&amp; index &lt; </span><span class="c0">this</span><span
                        class="c1">.data.length)<br> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c0">return</span><span class="c1">&nbsp;</span><span class="c0">this</span><span
                        class="c1">.data[index];<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c0">else</span><span
                        class="c1"><br> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c0">throw</span><span class="c1">&nbsp;</span><span class="c0">new</span><span
                        class="c1">&nbsp;</span><span class="c4">RangeError</span><span class="c1">(</span><span
                        class="c9">&quot;Index invalid&quot;</span><span class="c1">);<br> &nbsp; &nbsp;}<br> &nbsp;
                        &nbsp;<br> &nbsp; &nbsp;open(){<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;Baza de date a fost
                        deschis&#259;&quot;</span><span class="c1">);<br> &nbsp; &nbsp;}<br> &nbsp; &nbsp;<br> &nbsp;
                        &nbsp;close(){<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;Baza de date a fost
                        &icirc;nchis&#259;&quot;</span><span class="c1">);<br> &nbsp; &nbsp;}<br>}<br><br></span><span
                        class="c0">function</span><span class="c1">&nbsp;</span><span class="c4">get</span><span
                        class="c1">(index) {<br> &nbsp; &nbsp;</span><span class="c0">const</span><span
                        class="c1">&nbsp;db = </span><span class="c0">new</span><span class="c1">&nbsp;Database();<br>
                        &nbsp; &nbsp;db.open();<br> &nbsp; &nbsp;</span><span class="c0">try</span><span
                        class="c1">&nbsp;{<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c0">return</span><span
                        class="c1">&nbsp;db.getItem(index);<br> &nbsp; &nbsp;} </span><span class="c0">catch</span><span
                        class="c1">(err) {<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.error(err);<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c0">throw</span><span
                        class="c1">&nbsp;</span><span class="c0">new</span><span class="c1">&nbsp;</span><span
                        class="c4">Error</span><span class="c1">(err.message); &nbsp; </span><span class="c2">//
                        gener&#259;m din nou aceea&#537;i eroare</span><span class="c1"><br> &nbsp; &nbsp;} </span><span
                        class="c0">finally</span><span class="c1">&nbsp;{<br> &nbsp; &nbsp; &nbsp; &nbsp;db.close();<br>
                        &nbsp; &nbsp;}<br>}<br><br></span><span class="c0">function</span><span
                        class="c1">&nbsp;</span><span class="c4">printResult</span><span class="c1">(){<br> &nbsp;
                        &nbsp;</span><span class="c0">try</span><span class="c1">&nbsp;{<br> &nbsp; &nbsp; &nbsp;
                        &nbsp;</span><span class="c0">const</span><span class="c1">&nbsp;item = get(</span><span
                        class="c14">5</span><span class="c1">);<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(</span><span class="c9">&quot;Ob&#539;inut din
                        baza de date:&quot;</span><span class="c1">, item);<br> &nbsp; &nbsp;} </span><span
                        class="c0">catch</span><span class="c1">(err) {<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.log(err); &nbsp; </span><span class="c2">//
                        gestion&#259;m eroarea din func&#539;ia get</span><span class="c1"><br> &nbsp;
                        &nbsp;}<br>}<br><br>printResult();</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">&Icirc;n func&#539;ia get, dup&#259; gestionarea erorii, se genereaz&#259; din nou
            o eroare folosind operatorul throw:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.ffbe3a1622eceded6d50d2e16f515fcf50d5ec4a"></a><a
        id="t.7"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">try</span><span class="c1">&nbsp;{<br> &nbsp; &nbsp;</span><span
                        class="c0">return</span><span class="c1">&nbsp;db.getItem(index);<br>} </span><span
                        class="c0">catch</span><span class="c1">(err) {<br> &nbsp; &nbsp;</span><span
                        class="c4">console</span><span class="c1">.error(err);<br> &nbsp; &nbsp;</span><span
                        class="c0">throw</span><span class="c1">&nbsp;</span><span class="c0">new</span><span
                        class="c1">&nbsp;</span><span class="c4">Error</span><span
                        class="c1">(err.message);<br>}<br></span><span class="c0">finally</span><span class="c1">{<br>
                        &nbsp; &nbsp;db.close();<br>}</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">De asemenea, merit&#259; de men&#539;ionat c&#259; apelul db.close(), care
            &icirc;n mod condi&#539;ionat &icirc;nchide baza de date, este plasat &icirc;n blocul finally. Acest lucru
            garanteaz&#259; c&#259;, chiar dac&#259; se genereaz&#259; o eroare, opera&#539;ia respectiv&#259; va fi
            totu&#537;i executat&#259;.</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7"><span class="c8">Astfel, dac&#259; a avut loc o eroare la apelul db.getItem, atunci c&acirc;nd
            apel&#259;m func&#539;ia get, va avea loc, de asemenea, o eroare, astfel &icirc;nc&acirc;t putem gestiona
            aceast&#259; eroare &icirc;n func&#539;ia printResult:</span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p><a id="t.175d41030de6ff15e4c7ba7bc821fd07bf822ee8"></a><a
        id="t.8"></a>
    <table class="c12">
        <tr class="c13">
            <td class="c11" colspan="1" rowspan="1">
                <p class="c5"><span class="c0">function</span><span class="c1">&nbsp;</span><span
                        class="c4">printResult</span><span class="c1">(){<br> &nbsp; &nbsp;</span><span
                        class="c0">try</span><span class="c1">&nbsp;{<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span
                        class="c0">const</span><span class="c1">&nbsp;item = get(</span><span class="c14">5</span><span
                        class="c1">);<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(</span><span class="c9">&quot;Ob&#539;inut din baza de date:&quot;</span><span
                        class="c1">, item);<br> &nbsp; &nbsp;} </span><span class="c0">catch</span><span
                        class="c1">(err) {<br> &nbsp; &nbsp; &nbsp; &nbsp;</span><span class="c4">console</span><span
                        class="c1">.log(err);<br> &nbsp; &nbsp;}<br>}</span></p>
            </td>
        </tr>
    </table>
    <p class="c5 c7 c10"><span class="c8"></span></p>
    <p class="c5 c7 c10"><span class="c8"></span></p>
</body>

</html>